common/meta/types/src/user_auth.rs

```diff
@@ -11,17 +11,237 @@
+        match s {
+            PLAINTEXT_PASSWORD_STR => Ok(AuthType::PlaintextPassword),
+            SHA256_PASSWORD_STR => Ok(AuthType::Sha256Password),
+            DOUBLE_SHA1_PASSWORD_STR => Ok(AuthType::DoubleShaPassword),
+            NO_PASSWORD_STR => Ok(AuthType::NoPassword),
+            _ => Err(AuthType::bad_auth_types(s)),
+        }
+    }
+
+    pub fn to_str(&self) -> &str {
+        match self {
+            AuthType::NoPassword => NO_PASSWORD_STR,
+            AuthType::PlaintextPassword => PLAINTEXT_PASSWORD_STR,
+            AuthType::Sha256Password => SHA256_PASSWORD_STR,
+            AuthType::DoubleShaPassword => DOUBLE_SHA1_PASSWORD_STR,
+        }
+    }
+
+    fn bad_auth_types(s: &str) -> String {
+        let all = vec![
+            NO_PASSWORD_STR,
+            PLAINTEXT_PASSWORD_STR,
+            SHA256_PASSWORD_STR,
+            DOUBLE_SHA1_PASSWORD_STR,
+        ];
+        let all = all
+            .iter()
+            .map(|s| format!("'{}'", s))
+            .collect::<Vec<_>>()
+            .join("|");
+        format!("Expected auth type {}, found: {}", all, s)
+    }
+
+    fn get_password_type(self) -> Option<PasswordType> {
+        match self {
+            AuthType::PlaintextPassword => Some(PasswordType::PlainText),
+            AuthType::Sha256Password => Some(PasswordType::Sha256),
+            AuthType::DoubleShaPassword => Some(PasswordType::DoubleSha1),
+            _ => None,
+        }
+    }
+}
 
 #[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub enum AuthInfo {
+    None,
+    Password {
+        password: Vec<u8>,
+        password_type: PasswordType,
+    },
+}
+
+fn calc_sha1(v: &[u8]) -> [u8; 20] {
+    let mut m = ::sha1::Sha1::new();
+    m.update(v);
+    m.digest().bytes()
+}
+
+fn double_sha1(v: &[u8]) -> [u8; 20] {
+    calc_sha1(&calc_sha1(v)[..])
+}
+
+impl AuthInfo {
+    pub fn get_type(&self) -> AuthType {
+        match self {
+            AuthInfo::None => AuthType::NoPassword,
+            AuthInfo::Password {
+                password: _,
+                password_type: t,
+            } => match t {
+                PasswordType::PlainText => AuthType::PlaintextPassword,
+                PasswordType::Sha256 => AuthType::Sha256Password,
+                PasswordType::DoubleSha1 => AuthType::DoubleShaPassword,
+            },
+        }
+    }
+
+    pub fn get_password(&self) -> Option<Vec<u8>> {
+        match self {
+            AuthInfo::Password {
+                password: p,
+                password_type: _,
+            } => Some(p.to_vec()),
+            _ => None,
+        }
+    }
+
+    pub fn get_password_type(&self) -> Option<PasswordType> {
+        match self {
+            AuthInfo::Password {
+                password: _,
+                password_type: t,
+            } => Some(*t),
+            _ => None,
+        }
+    }
+
+    fn decode_double_sha1(
```

Comment:
`decode_double_sha1` is not a good name to me, it's confusing. How could we `decode` via hash? But I can't come up better names for it too... :cry: 

Do have any ideas?

Reply:
restore_sha1_mysql？